larq_compute_engine/core/bconv2d_impl_ref.h

```diff
@@ -0,0 +1,140 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+Modifications copyright (C) 2020 Larq Contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#ifndef COMPUTE_EGNINE_CORE_BCONV2d_IMPL_H_
+#define COMPUTE_EGNINE_CORE_BCONV2d_IMPL_H_
+
+#include "larq_compute_engine/core/bgemm_functor.h"
+#include "larq_compute_engine/core/packbits_utils.h"
+#include "tensorflow/lite/kernels/internal/common.h"
+#include "tensorflow/lite/kernels/internal/types.h"
+
+// This file is originally copied from
+// "tensorflow/lite/kernels/internal/reference/conv.h".
+// However, it's modified to perform binary convolution instead
+using namespace tflite;
+
+namespace compute_engine {
+namespace ce = compute_engine;
+namespace ref {
+
+template <class T, class TBitpacked>
+inline void BConv2D(const ConvParams& params, const RuntimeShape& input_shape,
+                    const T* input_data,
+                    const RuntimeShape& packed_filter_shape,
+                    const TBitpacked* packed_filter_data,
+                    const float* post_activation_multiplier_data,
+                    const float* post_activation_bias_data,
+                    const RuntimeShape& output_shape, T* output_data,
+                    const RuntimeShape& im2col_shape, T* im2col_data,
+                    bool bitpack_before_im2col, T* padding_buffer,
+                    const int pad_value, void* cpu_backend_context) {
+  using AccumScalar = std::int32_t;
+  using DstScalar = T;
+
+  const int stride_width = params.stride_width;
+  const int stride_height = params.stride_height;
+  const int dilation_width_factor = params.dilation_width_factor;
+  const int dilation_height_factor = params.dilation_height_factor;
+  const int pad_width = params.padding_values.width;
+  const int pad_height = params.padding_values.height;
+  // TODO(tom): for the backtransform we need the raw filter shape before
+  // bitpacking currently this is already passed as input argument. In the
+  // future, we will change this and this line needs to adapt.
+  // TODO: should the type be AccumScalar?
+  const std::int32_t backtransform_add = packed_filter_shape.Dims(1) *
+                                         packed_filter_shape.Dims(2) *
+                                         input_shape.Dims(3);
```

Comment:
Yes, the type here should be AccumScalar, but I'd rather leave it like this for now and fix it when we do the `backtransform_add` in a different way, in the int8 PR.